src/backend/vm_base/builder.rs

```diff
@@ -0,0 +1,198 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::shim::BootInfo;
+use super::*;
+
+use crate::binary::{Component, Segment};
+
+use kvm_ioctls::{Kvm, VmFd};
+use lset::Line;
+use mmarinus::{perms, Kind, Map};
+use nbytes::bytes;
+use primordial::Page;
+use x86_64::structures::paging::page_table::{PageTable, PageTableFlags};
+use x86_64::{align_up, PhysAddr, VirtAddr};
+
+use std::io::Result;
+use std::mem::size_of;
+use std::num::NonZeroUsize;
+
+pub struct Config {
+    pub n_syscall_pages: NonZeroUsize,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            n_syscall_pages: NonZeroUsize::new(256).unwrap(),
+        }
+    }
+}
+
+pub trait Hook {
+    fn shim_loaded(&mut self, _vm: &mut VmFd, _addr_space: &[u8]) -> Result<()> {
+        Ok(())
+    }
+
+    fn code_loaded(&mut self, _vm: &mut VmFd, _addr_space: &[u8]) -> Result<()> {
+        Ok(())
+    }
+
+    fn to_guest_phys(&self, addr: VirtAddr, start: VirtAddr) -> PhysAddr {
+        PhysAddr::new(addr.as_u64() - start.as_u64())
+    }
+}
+
+pub struct NoHook;
+impl Hook for NoHook {}
+
+pub struct Builder {
+    hook: Box<dyn Hook>,
+    shim: Component,
+    code: Component,
+    config: Config,
+}
```

Comment:
Do you need to own the hook? Or is it sufficient just to borrow it? Also, why dynamic dispatch here? Is this impossible?

```suggestion
pub struct Builder<T: Hook> {
    hook: T,
    shim: Component,
    code: Component,
    config: Config,
}
```